tests/Temporalio.Tests/Worker/WorkflowCodecHelperTests.cs

```diff
@@ -0,0 +1,171 @@
+namespace Temporalio.Tests.Worker;
+
+using Google.Protobuf;
+using Google.Protobuf.Collections;
+using Temporalio.Api.Common.V1;
+using Temporalio.Bridge.Api.WorkflowActivation;
+using Temporalio.Bridge.Api.WorkflowCompletion;
+using Temporalio.Worker;
+using Xunit;
+using Xunit.Abstractions;
+
+public class WorkflowCodecHelperTests : TestBase
+{
+    public WorkflowCodecHelperTests(ITestOutputHelper output)
+        : base(output)
+    {
+    }
+
+    [Fact]
+    public async Task CreateAndVisitPayload_Visiting_ReachesAllExpectedValues()
+    {
+        // This is just here to confirm our visitor works
+        var paths = new List<string>();
+        await CreateAndVisitPayload(
+            new(),
+            new WorkflowActivation(),
+            (ctx, _) =>
+            {
+                paths.Add(ctx.Path);
+                return Task.CompletedTask;
+            });
+        // Confirm some known paths are there for fixed field, repeated, and map
+        Assert.Contains("Jobs.StartWorkflow.Headers", paths);
+        Assert.Contains("Jobs.SignalWorkflow.Input", paths);
+        Assert.Contains("Jobs.ResolveActivity.Result.Completed.Result", paths);
+    }
+
+    [Fact]
+    public async Task EncodeAsync_AllPayloads_EncodesAll()
+    {
+        var comp = new WorkflowActivationCompletion();
+        var codec = new MarkerPayloadCodec();
+        await CreateAndVisitPayload(new(), comp, async (ctx, payload) =>
+        {
+            // We don't check search attributes on purpose
+            if (ctx.PropertyPath.Last().Item2 == "SearchAttributes")
+            {
+                return;
+            }
+            Assert.DoesNotContain("encoded", payload().Metadata.Keys);
+            await WorkflowCodecHelper.EncodeAsync(codec, comp);
+            if (!payload().Metadata.ContainsKey("encoded"))
+            {
+                Assert.Fail($"Payload at path {ctx.Path} not encoded");
+            }
+        });
+    }
+
+    [Fact]
+    public async Task DecodeAsync_AllPayloads_DecodesAll()
+    {
+        var act = new WorkflowActivation();
+        var codec = new MarkerPayloadCodec();
+        await CreateAndVisitPayload(new(), act, async (ctx, payload) =>
+        {
+            // We don't check search attributes on purpose
+            if (ctx.PropertyPath.Any(t => t.Item2 == "SearchAttributes"))
+            {
+                return;
+            }
+            Assert.DoesNotContain("decoded", payload().Metadata.Keys);
+            await WorkflowCodecHelper.DecodeAsync(codec, act);
+            if (!payload().Metadata.ContainsKey("decoded"))
+            {
+                Assert.Fail($"Payload at path {ctx.Path} not decoded");
+            }
+        });
+    }
+
+    // Creates payloads as needed, null context if already seen
+    private static async Task CreateAndVisitPayload(
```

Comment:
Nice!

Reply:
I knew you'd like this one (you may be the only one that will read/grok it). As expected, it helped me catch multiple issues.